detekt-rules/src/main/kotlin/org/example/detekt/PublicDataClassConstructorWithValueParameters.kt

```diff
@@ -0,0 +1,98 @@
+    try {
+      if (
+        ktClass.isData() &&
+          ktClass.hasPrimaryConstructor() &&
+          ktClass.primaryConstructor?.valueParameters?.any {
+            val ktpe = bindingContext.get(BindingContext.TYPE, it.typeReference)
+
+            (ktpe as KotlinType).unwrap().isValueClassType()
+          } == true
+      ) {
+        var hasJvmStaticConstructor = false
+        val hasCompanionObjects = ktClass.companionObjects.isNotEmpty()
+        if (!hasCompanionObjects) {
+          hasJvmStaticConstructor = false
+        } else {
+          ktClass.companionObjects.forEach { ktObjectDeclaration ->
+            if (!hasJvmStaticConstructor) {
+              ktObjectDeclaration.acceptChildren(
+                object : KtVisitorVoid() {
+                  override fun visitKtElement(element: KtElement) {
+                    try {
+                      element.acceptChildren(this)
+                    } catch (_: UnsupportedOperationException) {}
+                  }
+
+                  override fun visitNamedFunction(function: KtNamedFunction) {
+                    if (!hasJvmStaticConstructor) {
+                      val klassParametersAsString =
+                        ktClass.primaryConstructor?.valueParameterList?.parameters?.fold("") {
+                          acc,
+                          parameter ->
+                          if (acc == "") {
+                            acc + parameter.name + ":" + parameter.typeReference?.text
+                          } else {
+                            acc + ", " + parameter.name + ":" + parameter.typeReference?.text
+                          }
+                        }
+                      val functionParametersAsStrings =
+                        function.valueParameterList?.parameters?.fold("") { acc, parameter ->
+                          if (acc == "") {
+                            acc + parameter.name + ":" + parameter.typeReference?.text
+                          } else {
+                            acc + ", " + parameter.name + ":" + parameter.typeReference?.text
+                          }
+                        }
+                      val functionHasJvmStatic =
+                        function.annotationEntries.any { it.shortName?.asString() == "JvmStatic" }
+                      val functionHasAllButLastDefaultConstructorParameter =
+                        klassParametersAsString == functionParametersAsStrings
+                      hasJvmStaticConstructor =
+                        functionHasJvmStatic && functionHasAllButLastDefaultConstructorParameter
+                    }
+                  }
+                }
+              )
+            }
+          }
+        }
+        if (!hasJvmStaticConstructor) {
+          report(
+            CorrectableCodeSmell(
+              issue,
+              Entity.from(ktClass),
+              "Kotlin data classes must have a JVM static factory method to be compatible with other jvm languages.",
```

Comment:
I thought we discovered this was not the case if the constructor was properly annotated as `@JvmOverloads constructor`. Why do we need to enforce a static factory method?

Reply:
```suggestion
              "Kotlin data classes containing parameters that are value classes must have a JVM static factory method to be compatible with other jvm languages.",
```